Solution by Marceli P

The Iris dataset contains the Iris flower dataset of Fisher. It is included in the attachment. The file's class label is the last column in the file.
(Use the parameter skiprows = 1 to load the file)

1. On its basis, create a training and test set with parameter random_state = 0
2. With the help of the GridSearchCV function, find the best parameters for a logistic regression model from:
"C": 0.001, 0.01, 0.1, 1.0, 10.0, 100.0,
"penalty": 'l1', 'l2', 'elasticnet',
"solver": 'lbfgs', 'liblinear', 'sag'

(some parameters do not work together and you will be warned about this)

3. Enter the parameters of the best classifier.
4. Using a trained classifier, make a prediction on the test set and enter the average prediction accuracy.
5. Use StandardScaler () and normalize the dataset.
6. As in step 2, train a new classifier on the normalized data.
7. Using the second trained classifier, make a prediction on the test set and enter the average prediction accuracy.
8. Enter the parameters of the best, new classifier

In [95]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV 
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import average_precision_score

In [96]:
df = pd.read_csv('iris.csv', sep=',')

In [97]:
df.head()

sepal.length  sepal.width  petal.length  petal.width  variety
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

In [98]:
X = df.iloc[:, :3]
y = df.iloc[:, 4]
#print(X,y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [99]:
logreg = LogisticRegression()

grid_all = {"C": [0.001, 0.01, 0.1, 1.0, 10.0, 100.0],
            "penalty": ['l1', 'l2', 'elasticnet'],
            "solver": ['lbfgs', 'liblinear', 'sag']}




logreg_cv = GridSearchCV(logreg, grid_all, cv=10)
logreg_cv.fit(X_train,y_train)

print("Tuned hyperparameters :(the best parameters) " , logreg_cv.best_params_)
print("Best score :" , logreg_cv.best_score_)

C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: Conve

C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear 

Tuned hyperparameters :(the best parameters)  {'C': 1.0, 'penalty': 'l2', 'solver': 'sag'}
Best score : 0.9800000000000001


C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\

Tuned hyperparameters :(the best parameters)  {'C': 1.0, 'penalty': 'l2', 'solver': 'sag'}

Best score : 0.9800000000000001

4. Using a trained classifier, make a prediction on the test set and enter the average prediction accuracy.

In [100]:
y_true, y_pred = y_test, logreg_cv.predict(X_test)


print(classification_report(y_true, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       0.94      0.89      0.91        18
           2       0.83      0.91      0.87        11

    accuracy                           0.93        45
   macro avg       0.92      0.93      0.93        45
weighted avg       0.94      0.93      0.93        45



In [101]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import accuracy_score

print("Balanced accuracy score :", balanced_accuracy_score(y_true, y_pred))
print ("Average accuracy is: ", accuracy_score(y_true, y_pred) * 100, "%")

Balanced accuracy score : 0.9326599326599326
Average accuracy is:  93.33333333333333 %


5. Use StandardScaler () and normalize the dataset.

In [102]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_x = scaler.fit_transform(X)

#print(scale_x)

6. As in step 2, train a new classifier on the normalized data.

In [103]:
X_train, X_test, y_train, y_test = train_test_split(scaled_x, y, test_size=0.3, random_state=0)

logreg = LogisticRegression()

grid_all = {"C": [0.001, 0.01, 0.1, 1.0, 10.0, 100.0],
            "penalty": ['l1', 'l2', 'elasticnet'],
            "solver": ['lbfgs', 'liblinear', 'sag']}




logreg_cv = GridSearchCV(logreg, grid_all, cv=10)
logreg_cv.fit(X_train,y_train)

print("Tuned hyperparameters :(the best parameters) " , logreg_cv.best_params_)
print("Best score :" , logreg_cv.best_score_)



Tuned hyperparameters :(the best parameters)  {'C': 10.0, 'penalty': 'l2', 'solver': 'lbfgs'}
Best score : 0.9700000000000001


C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\Marceli\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\sklearn\linear_model\_sag.py:352: Conve

Previous:

* Tuned hyperparameters :(the best parameters)  {'C': 1.0, 'penalty': 'l2', 'solver': 'sag'}
* Accuracy : 0.9800000000000001




With normalized dataset:

* Tuned hyperparameters :(the best parameters)  {'C': 10.0, 'penalty': 'l2', 'solver': 'lbfgs'}
* Accuracy : 0.9700000000000001

In [104]:
from sklearn.metrics import balanced_accuracy_score

y_true, y_pred = y_test, logreg_cv.predict(X_test)
print("Balanced accuracy score :", balanced_accuracy_score(y_true, y_pred))
print ("Average accuracy is: ", accuracy_score(y_true, y_pred) * 100, "%")

Balanced accuracy score : 0.9208754208754208
Average accuracy is:  93.33333333333333 %


Results without a normalized dataset have better predictions by about 1‰